### DISCUSSION TITLE
Composite rows with independent active item index

### DISCUSSION THREAD

**Discussion by gnapse** [ariakit/ariakit#3905] [Questions] [answered]
I have a situation where a composite with rows would fit, but with a twist.

Composites with rows work like a grid: if you move to the 4th index in the first row, and then press the down arrow, you land in the 4th item of the 2nd row.

I need to implement a situation in which moving to a new row with the up or down arrow will go to the 1st item of the previous or next row, or pick up where it left off if that row has been visited before.

Is that possible?

**Comment by diegohaz** [comment, marked as answer]
Could you better describe your specific use case?

You will need to implement custom logic within the function form of the [`moveOnKeyPress`](https://ariakit.org/reference/composite-item#moveonkeypress) prop.

**Comment by gnapse** [comment]
Imagine a UI like this:

```
First section
-----------------------------
A1 A2 A3 A4 A5 A6 A7 A8 A9
-----------------------------

Second section
-----------------------------
B1 B2 B3 B4 B5
-----------------------------
```

There may be a few more sections like this. Each section has an independent number of items inside it. The items are listed horizontally, inside a potentially scrollable container (denoted by the horizontal lines). Each scrollable container is independent of the other ones in other sections.

Still, I thought it could be nice if left/right arrows navigate through a section, and that the up/down arrow navigate across sections. But when I go to a new section, I do not want to land in the same item index in which I'm at. If I go right to A6 and then press down, I want to land in B1.

Perhaps I'll avoid it for now. I managed to find an acceptable alternative: each section is a horizontally-oriented composite. And pressing tab or shift-tab takes you to the next or previous sections. I'll leave this UI like that right now, but I may explore customizing `moveOnKeyPress` in the future if I really want this.

Thanks!

**Reply by diegohaz** [reply]
> Perhaps I'll avoid it for now. I managed to find an acceptable alternative: each section is a horizontally-oriented composite. And pressing tab or shift-tab takes you to the next or previous sections. I'll leave this UI like that right now, but I may explore customizing `moveOnKeyPress` in the future if I really want this.

Yes, that sounds better.

The composite docs aren't finished, but it's important to note: composite widgets need a specific composite role (listbox, toolbar, menu, tablist, etc.) to function properly with screen readers. Without this, users won't be able to navigate between items using arrow keys.

By default, Windows screen readers operate in browse mode, capturing arrow key presses for their own commands (e.g., horizontal arrow keys read the previous/next character in a text/paragraph). However, with the correct composite role, screen readers will automatically switch to focus mode and delegate all keystrokes to JavaScript, allowing users to navigate with arrow keys.

As a last resort, you can also enable tabbing between composite items if it's not clear to users that they should use arrow keys: [https://ariakit.org/reference/composite-item#tabbable](https://ariakit.org/reference/composite-item#tabbable).
